odbc/src/api/data.rs

```diff
@@ -443,12 +643,59 @@ mod i32_len {
     ///
     pub unsafe fn set_output_string(
         message: &str,
+        col_num: USmallInt,
+        var_data_indicies: &mut HashMap<USmallInt, (usize, usize)>,
         output_ptr: *mut Char,
         buffer_len: usize,
         text_length_ptr: *mut Len,
     ) -> SqlReturn {
-        let (len, ret) = set_output_string_helper(message, output_ptr, buffer_len);
+        let index = if let Some((index, _)) = var_data_indicies.get(&col_num) {
+            *index
+        } else {
+            0usize
+        };
+        if index >= message.len() {
+            *text_length_ptr = 0;
+            return SqlReturn::NO_DATA;
+        }
+        let (len, ret) =
+            set_output_string_helper(message.get(index..).unwrap(), output_ptr, buffer_len);
+        // the returned length should always be the total length of the data.
+        *text_length_ptr = message.chars().count() as Len;
+        var_data_indicies.insert(col_num, (len + index, 0usize));
+        ret
+    }
+
+    ///
+    /// set_output_binary writes [`message`] to the *Char [`output_ptr`]. [`buffer_len`] is the
+    /// length of the [`output_ptr`] buffer in characters; the message should be truncated
+    /// if it is longer than the buffer length. The number of characters written to [`output_ptr`]
+    /// should be stored in [`text_length_ptr`].
+    ///
+    /// # Safety
+    /// This writes to multiple raw C-pointers
+    ///
+    pub unsafe fn set_output_binary(
```

Comment:
Same for this. It should support streaming binary data.
This is not high priority though.
But since you have a framework in place which let's you know the offset for each column and total len, it seem that you could apply it here too.
I also think that having the share code/logic which apply to all variable length data type would make think easier to maintain later on.
The part where you retrieve the index and decide if things should stop here returning NO_DATA is the same, and the part updating the internal offsets too.

Reply:
It currently works. The code is just more confusing than it needs to be.